qiling/debugger/qdb/render/render.py

```diff
@@ -0,0 +1,233 @@
+        """
+
+        lines = ""
+        for idx, r in enumerate(regs, 1):
+            line = "{}{}: 0x{{:08x}}  {}\t".format(COLORS[(idx-1) // self.regs_a_row], r, color.END)
+
+            if diff_reg and r in diff_reg:
+                line = f"{color.UNDERLINE}{color.BOLD}{line}"
+
+            if idx % self.regs_a_row == 0 and idx != 32:
+                line += "\n"
+
+            lines += line
+
+        print(lines.format(*regs.values()))
+
+    def render_stack_dump(self, arch_sp: int, pointer_size: int = 4) -> None:
+        """
+        helper function for redering stack dump
+        """
+
+        for idx in range(self.stack_num):
+            addr = arch_sp + idx * pointer_size
+            if (val := self.try_read_pointer(addr)[0]):
+                print(f"$sp+0x{idx*pointer_size:02x}│ [0x{addr:08x}] —▸ 0x{self.unpack(val):08x}", end="")
+
+            # try to dereference wether it's a pointer
+            if (buf := self.try_read_pointer(addr))[0] is not None:
+
+                if (addr := self.ql.unpack(buf[0])):
+
+                    # try to dereference again
+                    if (buf := self.try_read_pointer(addr))[0] is not None:
+                        s = self.try_read_string(addr)
+
+                        if s and s.isprintable():
+                            print(f" ◂— {self.read_string(addr)}", end="")
+                        else:
+                            print(f" ◂— 0x{self.unpack(buf[0]):08x}", end="")
+            print()
+
+    def render_assembly(self, lines) -> None:
+        """
+        helper function for rendering assembly
+        """
+
+        # assembly before current location
+        if (backward := lines.get("backward", None)):
+            for line in backward:
+                self.print_asm(line)
+
+        # assembly for current location
+        if (cur_insn := lines.get("current", None)):
+            prophecy = self.predictor.predict()
+            self.print_asm(cur_insn, to_jump=prophecy.going)
+
+        # assembly after current location
+        if (forward := lines.get("forward", None)):
+            for line in forward:
+                self.print_asm(line)
+
+    def swap_reg_name(self, cur_regs: Mapping["str", int], extra_dict=None):
+        """
+        swap register name with more readable register name
+        """
+
+        target_items = extra_dict.items() if extra_dict else self.regs_need_swaped.items()
+
+        for old_reg, new_reg in target_items:
+            cur_regs.update({old_reg: cur_regs.pop(new_reg)})
+
+        return cur_regs 
+
+    def print_asm(self, insn: CsInsn, to_jump: bool = False) -> None:
+        """
+        helper function for printing assembly instructions, indicates where we are and the branch prediction
+        provided by BranchPredictor
+        """
+
+        opcode = "".join(f"{b:02x}" for b in insn.bytes)
+        trace_line = f"0x{insn.address:08x} │ {opcode:15s} {insn.mnemonic:10} {insn.op_str:35s}"
```

Comment:
Assembly line format could differ quite a lot between architectures.
Consider having these values as inherited class members that change from arch to arch.